polyply/src/annotate_dna.py

```diff
@@ -0,0 +1,194 @@
+        return strands
+
+    def _determine_sequence(self, strand):
+        sequence = []
+        if nx.cycle_basis(strand):
+            start_base = sample(strand.nodes, 1)
+        else:
+            condition = lambda node: strand.degree(node) == 1
+            start_base = next(filter(condition, strand.nodes), None)
+        for resid in nx.dfs_tree(strand, start_base):
+            sequence.append(strand.nodes[resid]['resname'])
+        return sequence
+
+    def _find_complementary_strands(self, ref_strand, strands):
+        # Sequence we want to find match for
+        ref_seq = self._determine_sequence(ref_strand)
+
+        match_ratios = {}
+        for ndx, strand in enumerate(strands, start=1):
+            seq = self._determine_sequence(strand)
+
+            match_ratio = 0
+            for ref_base, base in zip(ref_seq, seq):
+                if base_library[ref_base] == base:
+                    match_ratio += 1
+            match_ratios[ndx] = match_ratio/len(ref_seq)
+
+            match_ratio = 0
+            for ref_base, base in zip(ref_seq, reversed(seq)):
+                if base_library[ref_base] == base:
+                    match_ratio += 1
+            match_ratios[-ndx] = match_ratio/len(ref_seq)
+
+        match =  max(match_ratios, key=match_ratios.get)
+        match_strand = strands.pop(abs(match) - 1)
+
+        if match <= 0:
+            mapping = dict(zip(match_strand, sorted(match_strand, reverse=True)))
+            match_strand = nx.relabel_nodes(match_strand, mapping)
+
+        return match_strand
+
+    def _combine_complementary_strands(self, meta_molecule, strands):
+        """
+        Given a DNA molecule generate the appropriate meta_molecule used
+        for the coordinate generation and backmapping. The constructed
+        meta_molecule overwrites the one generated by polyply.
+
+        Parameters
+        ------------
+        meta_molecule: :class:`polyply.src.MetaMolecule`
+        strands: list[:class: `networkx.Graph`]
+        """
+
+        while len(strands) > 1:
+            ref_strand = strands.pop()
+
+            match_strand = self._find_complementary_strands(ref_strand, strands)
+
+            for findex, bindex in zip(ref_strand, sorted(match_strand)):
+
+                fbase = ref_strand.nodes[findex]
+                bbase = match_strand.nodes[bindex]
+
+                current_node = fbase
+
+                # Combine attribute of strands
+                current_node['resid'] = findex
+                current_node['build'] = fbase['build'] | bbase['build']
+                current_node['nnodes'] = fbase['nnodes'] + bbase['nnodes']
+                current_node['nedges'] = fbase['nedges'] + bbase['nedges']
+                current_node['graph'] = nx.compose(fbase['graph'], bbase['graph'])
+                current_node['density'] = (fbase['density'] + bbase['density']) / 2
+                current_node['resname'] = f"{fbase['resname']},{bbase['resname']}"
```

Comment:
Same as before put in loop to reduce lines in code; also make the attribute list a class argument so it could be changed later

Reply:
I accidentally resolved the conversation where I previously mentioned this. Not really sure how to write the for-loop so that it reduces the total number of lines. Do you suggest  a dictionary, since just adding the old components works for only three cases, maybe I am overlooking something obvious?

```python
new_attributes = {'resid': findex, 'graph': nx.compose(...), ...}
for attribute in attribute_list:
    current_node[attribute] = new_attributes[attribute]
```